django_fido/forms.py

```diff
@@ -89,3 +93,52 @@ def clean_signature(self) -> bytes:
             return base64.b64decode(value)
         except ValueError:
             raise ValidationError(_('FIDO 2 response is malformed.'), code='invalid')
+
+
+class Fido2ModelAuthenticationForm(AuthenticationForm, Fido2AuthenticationForm):
+    """Authentication form with username, password and FIDO 2 credentials."""
+
+    def __init__(self, request: HttpRequest, fido2_server: Fido2Server, session_key: str, *args, **kwargs):
+        """Initialize form."""
+        super().__init__(request, *args, **kwargs)
+        self.fido2_server = fido2_server
+        self.session_key = session_key
+        self.error_messages['invalid_login'] = _(
+            "Please enter a correct %(username)s and password and use valid FIDO2 security key."
+        )
+
+    def clean(self):
+        """Authenticate user."""
+        username = self.cleaned_data.get('username')
+        password = self.cleaned_data.get('password')
+        state = self.request.session.pop(self.session_key, None)
+        if state is None:
+            raise ValidationError(_('Authentication request not found.'), code='missing')
+
+        if username is not None and password:
+            self.user_cache = authenticate(
+                self.request,
+                username=username,
+                password=password,
+                fido2_server=self.fido2_server,
+                fido2_state=state,
+                fido2_response=self.cleaned_data,
+            )
+            if self.user_cache is None:
+                raise self.get_invalid_login_error()
+            else:
+                self.confirm_login_allowed(self.user_cache)
+        else:
+            raise self.get_invalid_login_error()
```

Comment:
This is not covered by tests.

Reply:
I updated this case as discussed. Check for username and password is unnecessary, because they are required fields.